services/controllerhost/dfdd.go

```diff
@@ -58,246 +62,292 @@ type (
+		delete(copy, event.Key)
+		failedEvent = NewInputHostFailedEvent(event.Key)
+	case common.StoreServiceName:
+		dfddHost := newDFDDHost(dfddHostStateDown, dfdd.timeSource)
+		copy[event.Key] = dfddHost
+		failedEvent = NewStoreHostFailedEvent(event.Key)
+	}
+
+	dfdd.putHosts(service, copy)
+	if !dfdd.context.eventPipeline.Add(failedEvent) {
+		dfdd.context.log.WithField(common.TagEvent, event).Error("failed to enqueue event")
 	}
 }
 
-func (dfdd *dfddImpl) handleListenerEvent(id serviceID, listenerEvent *common.RingpopListenerEvent) {
+func (dfdd *dfddImpl) handleHostGoingDownEvent(service string, event *common.RingpopListenerEvent) {
+
+	hosts, err := dfdd.getHosts(service)
+	if err != nil {
+		return
+	}
 
-	if listenerEvent.Type == common.HostAddedEvent {
-		dfdd.reportHostHealthy(id, listenerEvent.Key)
+	curr, ok := hosts[event.Key]
+	if !ok || curr.state != dfddHostStateUP {
 		return
 	}
 
-	dfdd.reportHostUnhealthy(id, listenerEvent.Key)
+	copy := deepCopyMap(hosts)
+	copy[event.Key] = newDFDDHost(dfddHostStateGoingDown, dfdd.timeSource)
+	dfdd.putHosts(service, copy)
+}
+
+func (dfdd *dfddImpl) handleListenerEvent(service string, event *common.RingpopListenerEvent) {
+	switch event.Type {
+	case common.HostAddedEvent:
+		dfdd.handleHostAddedEvent(service, event)
+	case common.HostRemovedEvent:
+		dfdd.handleHostRemovedEvent(service, event)
+	case hostGoingDownEvent:
+		dfdd.handleHostGoingDownEvent(service, event)
+	}
+}
 
-	var event Event
+func (dfdd *dfddImpl) getHosts(service string) (map[string]dfddHost, error) {
+	switch service {
+	case common.InputServiceName:
+		return dfdd.inputHosts.Load().(map[string]dfddHost), nil
+	case common.StoreServiceName:
+		return dfdd.storeHosts.Load().(map[string]dfddHost), nil
+	default:
+		return nil, errUnknownService
+	}
+}
 
-	switch id {
-	case inputServiceID:
-		dfdd.context.log.WithField(common.TagIn, common.FmtIn(listenerEvent.Key)).Info("InputHostFailed")
-		event = NewInputHostFailedEvent(listenerEvent.Key)
-	case storeServiceID:
-		dfdd.context.log.WithField(common.TagStor, common.FmtStor(listenerEvent.Key)).Info("StoreHostFailed")
-		event = NewStoreHostFailedEvent(listenerEvent.Key, hostDownStage1)
+func (dfdd *dfddImpl) putHosts(service string, hosts map[string]dfddHost) {
+	switch service {
+	case common.InputServiceName:
+		dfdd.inputHosts.Store(hosts)
+	case common.StoreServiceName:
+		dfdd.storeHosts.Store(hosts)
 	default:
-		dfdd.context.log.Error("ListenerEvent for unknown service")
 		return
 	}
+}
+
+func newDFDDHost(state dfddHostState, timeSource common.TimeSource) dfddHost {
+	return dfddHost{
+		state:               state,
+		lastStateChangeTime: timeSource.Now().UnixNano(),
+	}
+}
 
-	if !dfdd.context.eventPipeline.Add(event) {
-		dfdd.context.log.WithField(common.TagEvent, event).Error("Failed to enqueue event")
+func deepCopyMap(src map[string]dfddHost) map[string]dfddHost {
+	copy := make(map[string]dfddHost, common.MaxInt(len(src), 8))
```

Comment:
hmm.. we call this method from HostAddedEvent. So chances are if we allocate it to the same size as before, the map will be reallocted when we add an entry. That's why its a maxInt. If you are asking this question, it means the intention isn't clear from this code. I will remove the maxInt and pass along the desiredNewCapacity as a param or just simply leave not optimize it.